### PR TITLE
Js pipes

### PR DESCRIPTION
**Pull request by glacambre**
Issue with previous discussion: https://github.com/cmcaine/tridactyl/issues/563

My code doesn't have the error handling the previous code had, but I'm not sure passing errors through pipes is a good idea. I believe it's better to just let the whole pipeline break. Are there disadvantages to not doing any "internal" error handling?

Also, I've added an "arg" variable to js and jsb for easy access to the piped value. I thought a bit about how to automatically insert "//" in the evaluated string but I can't come up with something that would be reliable. We could add a flag to js, perhaps "-p" for "pipe", which would be required to be at the very beginning of the command and would let js know it should insert "//" at the end of the string. Not having to manually remove the "//" at the end of a command you're editing from your history could be nice.
Are there alternatives I haven't thought of?

### CODE REVIEW

File: src/excmds.ts [hunk 1]
```diff
@@ -2425,6 +2427,7 @@ export async function echo(...str: string[]) {
  */
```
Comment (bovine3dom) [inline src/excmds.ts]: Please document `arg` :)

File: src/excmds.ts [hunk 2]
```diff
@@ -2425,6 +2427,7 @@ export async function echo(...str: string[]) {
...
 //#content
 export async function js(...str: string[]) {
+    let arg = str[str.length - 1]
```
Comment (bovine3dom) [inline src/excmds.ts]: Please call this something that makes it clear it is magic, e.g. `JS_ARG`.

File: src/excmds.ts (hunk 1 above)
Comment (bovine3dom) [reply src/excmds.ts]: You should also document the `//` trick for allowing it to receive an argument from `composite`.

@cmcaine: I'd like to hear what you think about `//`.

**Comment by bovine3dom** [comment]
I think `js -p` could be a sound idea - outside of a pipe, `arg` will probably be nonsense, so we shouldn't define it at all.

**Comment by glacambre** [comment]
Yup, I tested this on several examples:
```
composite get_current_url | fillcmdline_notrail
composite js [1, 2, 3] | js -p alert(JS_ARG[0] + " " + JS_ARG[3])
composite jsb (x) => console.log("hello " + x) | jsb -p JS_ARG("world")
```
The one thing that doesn't work is piping functions to `js` because we can't send functions from the background script to the content script. However, we might be able to work around that by stringifying functions and re-evaluating them on the other side.

File: src/excmds.ts (hunk 1 above)
Comment (glacambre) [reply src/excmds.ts]: I've implemented -p now, which means we don't need the `//` hack anymore.

**Comment by bovine3dom** [comment]
For me, this breaks `bind D composite tabprev; sleep 100; tabclose #`. It closes two tabs rather than just one, and seems to close tabs that aren't the tab that should be closed.

Could you look into it, @glacambre?

**Comment by glacambre** [comment]
`composite tabprev; sleep 100; tabclose #` is a nasty hack and I'm glad my PR breaks it >:(

I think the best way to fix this is to implement a mechanism that ensures that each command has finished executing before the next one is executed. It might take me a little while but I'll try to do that.

**Comment by bovine3dom** [comment]
Oh, if that's it, just plonk `await` in front of it :)

File: src/excmds.ts [hunk 3]
```diff
@@ -1657,23 +1658,24 @@ export function repeat(n = 1, ...exstr: string[]) {
...
+    try {
+        return cmds
+            .join(" ")
+            .split(";")
+            .reduce(
+                async (_, cmd) => {
+                    return cmd.split("|").reduce(
+                        async (pipedValue, cmd) => {
+                            let [fn, args] = excmd_parser.parser(cmd)
+                            return fn.call({}, ...args, await pipedValue)
```
Comment (bovine3dom) [inline src/excmds.ts]: I reckon shove an await here and you'll be golden.

**Comment by glacambre** [comment]
There actually were multiple issues: the first was indeed a missing await and the second one was that I assumed passing an additional undefined argument to the first function of a pipeline would be harmless while it wasn't.
Both issues are now fixed.

**Comment by bovine3dom** [comment]
I realise this is a bit late now, but `-p` isn't much to do with pipes, does it?

e.g. `command annoyme js -p alert(JS_ARG)` -> `annoyme "annoying"`. I guess the gotcha if you don't use echo and composite is that you can't have a space in the argument?

**Pull request merged**